### DISCUSSION TITLE
How exacalidraw's pen could draw a random user-drawn curve smoothly on the fly

### DISCUSSION THREAD

**Discussion by ashuvssut** [excalidraw/excalidraw#3362] [Q&A] [answered]
with a stroke roughness of 2 i was able to draw a continuous  rough line [here](https://deploy-preview-791--doc2pen.netlify.app/sketch) in my project
![image](https://user-images.githubusercontent.com/60546840/112832858-c3175d00-904a-11eb-9413-f708741cedb2.png)

The method I followed to draw a line was.....
when the you want to draw a line on the canvas... first I record the coordinates of the points your mouse moved in an array
```js
let pathArray = []
pathArray.push( [point.x, point.y] )
.
.
.
.
//after recording all points... draw the curve
roughCanvas.curve(penPath, { strokeWidth: width, roughness: roughness });
```

**But this has an obivious demerit:-**  User when draws a line, for say, 3seconds.... in those three seconds the coordinates will be recorded and nothing will be drawn on the canvas. after those 3 seconds that curve is drawn. This is totally undesired. i must draw those on the fly

But excalidraw draws those on the fly. How did you achieve that?

**Comment by dwelle** [comment, marked as answer]
Just add the intermediate points and render continually. I suppose you're doing that already when drawing other shape types.

**Reply by ashuvssut** [reply]
@dwelle, thanks for replying

Previous I had 3 eventlisteners for that canvas
```js

function handleMouseDown(point) {
  penPath.push([point.x, point.y]);
}

function handleMouseMove(point) {//executes multiple time (whenever the mouse is moved with 'drawing mode' enabled)
  penPath.push([point.x, point.y]);
}

function handleMouseUp(event) {
  roughCanvas.curve(penPath, { strokeWidth: width, roughness: roughness });
  penPath = []; //reset array
}
```
**RESULT**
![image](https://user-images.githubusercontent.com/60546840/112832858-c3175d00-904a-11eb-9413-f708741cedb2.png)

--------------------------------------------------------------

After your suggestion I came up with some changes... It worked fine but still needs one improvement. I want your help with that

### CHANGE
changes in 2nd eventlistener
```js
const [typeState, setTypeState] = useState( 
  context.getImageData(0, 0, canvasWidth, canvasHeight)
); //<<<   new changes

function handleMouseDown(point) {
  penPath.push([point.x, point.y]);
}

function handleMouseMove(point) {//executes multiple time (whenever the mouse is moved with 'drawing mode' enabled)
  context.putImageData(typeState, 0, 0); //<<<   new changes
  penPath.push([point.x, point.y]); 
  roughCanvas.curve(penPath, { strokeWidth: width, roughness: roughness }); //<<<   new changes
}

function handleMouseUp(event) {
  roughCanvas.curve(penPath, { strokeWidth: width, roughness: roughness });
  penPath = [];
}
```
**worked perfectly fine**

**BUT**

You excalidraw pen draws in a nice manner

**MY PEN** ---------TRY it out here:- https://deploy-preview-816--doc2pen.netlify.app/sketch
![my pen](https://user-images.githubusercontent.com/60546840/113053423-1da0de00-915d-11eb-8676-02891c228d1a.gif)

**EXCALIDRAW's PEN**
![excalidraw](https://user-images.githubusercontent.com/60546840/113053481-28f40980-915d-11eb-9d7f-6e164467347e.gif)


How excalidraw's pen draws in such a nice smooth non-wiggly manner?

**Reply by ashuvssut** [reply]
@dwelle, Please suggest what changes should I make so that I can get my desired result

I ud be glad if you could give me hints by some code
OR
you can also give me the file name that I should look in you excalidraw repo to get some hints


thanks in advance
